sw/otbn/crypto/p384_curve_point_valid.s

```diff
@@ -0,0 +1,204 @@
+/* Copyright lowRISC contributors. */
+/* Licensed under the Apache License, Version 2.0, see LICENSE for details. */
+/* SPDX-License-Identifier: Apache-2.0 */
+
+/**
+ * Curve point validation for curve P-384.
+ *
+ * This binary has the following modes of operation:
+ * 1. MODE_SHARED_KEYGEN: check if a given public key for ECDH shared key
+ *                        generation is a valid curve point
+ */
+
+ /**
+ * Mode magic values generated with
+ * $ ./util/design/sparse-fsm-encode.py -d 6 -m 4 -n 11 \
+ *    --avoid-zero -s 3660400884
+ *
+ * Call the same utility with the same arguments and a higher -m to generate
+ * additional value(s) without changing the others or sacrificing mutual HD.
+ *
+ * TODO(#17727): in some places the OTBN assembler support for .equ directives
+ * is lacking, so they cannot be used in bignum instructions or pseudo-ops such
+ * as `li`. If support is added, we could use 32-bit values here instead of
+ * 11-bit.
+ */
+.equ MODE_SHARED_KEY, 0x5ec
+
+.section .text.start
+start:
+  /* Init all-zero register. */
+  bn.xor    w31, w31, w31
+
+  /* Read the mode and tail-call the requested operation. */
+  la        x2, mode
+  lw        x2, 0(x2)
+
+  addi      x3, x0, MODE_SHARED_KEY
+  beq       x2, x3, p384_curve_point_valid
+
+  /* Unsupported mode; fail. */
+  unimp
+  unimp
+  unimp
+
+/**
+ * Check if a provided curve point is valid.
+ *
+ * For a given curve point (x, y), check that:
+ * - x and y are both fully reduced mod p
+ * - (x, y) is on the P-384 curve.
+ *
+ * Note that, because the point is in affine form, it is not possible that (x,
+ * y) is the point at infinity. In some other forms such as projective
+ * coordinates, we would need to check for this also.
+ *
+ * This routine raises a software error and halts operation if the curve point
+ * is invalid.
+ *
+ * @param[in]  dmem[12]: dptr_r, pointer to dmem location where right
+ *                               side result r will be stored
+ * @param[in]  dmem[16]: dptr_s, pointer to dmem location where left side
+ *                               result s will be stored
+ * @param[in]  dmem[20]: dptr_x, pointer to dmem location containing affine
+ *                               x-coordinate of input point
+ * @param[in]  dmem[24]: dptr_y, pointer to dmem location containing affine
+ *                               y-coordinate of input point
+ *
+ * Flags: Flags have no meaning beyond the scope of this subroutine.
+ *
+ * clobbered registers: x2, x3, x20 to x23, w0 to w17
+ * clobbered flag groups: FG0
+ */
+p384_curve_point_valid:
```

Comment:
Would it make sense to move this to `p384_isoncurve.s` so it doesn't have to be repeated in both the test and the entrypoint?

Reply:
I thought of that too, and don't know why I decided to not do it. :D 
But you're right, should make the code clearer.